modules/network_synchronizer/scene_synchronizer.cpp

```diff
@@ -0,0 +1,1845 @@
+		return;
+	}
+
+	// Find the best recoverable input_id.
+	uint32_t checkable_input_id = UINT32_MAX;
+	if (player_controller->last_known_input() != UINT32_MAX && player_controller->get_stored_input_id(-1) != UINT32_MAX) {
+		// Find the best snapshot to recover from the one already
+		// processed.
+		if (client_snapshots.empty() == false) {
+			for (
+					auto s_snap = server_snapshots.rbegin();
+					checkable_input_id == UINT32_MAX && s_snap != client_snapshots.rend();
+					++s_snap) {
+				for (auto c_snap = client_snapshots.begin(); c_snap != client_snapshots.end(); ++c_snap) {
+					if (c_snap->input_id == s_snap->input_id) {
+						// Server snapshot also found on client, can be checked.
+						checkable_input_id = c_snap->input_id;
+						break;
+					}
+				}
+			}
+		}
+	}
+
+	if (checkable_input_id == UINT32_MAX) {
+		// No snapshot found, nothing to do.
+		return;
+	}
+
+#ifdef DEBUG_ENABLED
+	// Unreachable cause the above check
+	CRASH_COND(server_snapshots.empty());
+	CRASH_COND(client_snapshots.empty());
+#endif
+
+	// Drop all the old server snapshots until the one that we need.
+	while (server_snapshots.front().input_id < checkable_input_id) {
+		server_snapshots.pop_front();
+	}
+
+	// Drop all the old client snapshots until the one that we need.
+	while (client_snapshots.front().input_id < checkable_input_id) {
+		client_snapshots.pop_front();
+	}
+
+#ifdef DEBUG_ENABLED
+	// These are unreachable at this point.
+	CRASH_COND(server_snapshots.empty());
+	CRASH_COND(server_snapshots.front().input_id != checkable_input_id);
+
+	// This is unreachable, because we store all the client shapshots
+	// each time a new input is processed. Since the `checkable_input_id`
+	// is taken by reading the processed doll inputs, it's guaranteed
+	// that here the snapshot exists.
+	CRASH_COND(client_snapshots.empty());
+	CRASH_COND(client_snapshots.front().input_id != checkable_input_id);
+#endif
+
+	// --- Phase two: compare the server snapshot with the client snapshot. ---
+	bool need_recover = false;
+	bool recover_controller = false;
+	LocalVector<SceneSynchronizer::NodeData *> nodes_to_recover;
+	LocalVector<SceneSynchronizer::PostponedRecover> postponed_recover;
+
+	nodes_to_recover.reserve(server_snapshots.front().node_vars.get_num_elements());
+	for (
+			OAHashMap<ObjectID, Vector<SceneSynchronizer::VarData>>::Iterator s_snap_it = server_snapshots.front().node_vars.iter();
+			s_snap_it.valid;
+			s_snap_it = server_snapshots.front().node_vars.next_iter(s_snap_it)) {
+		SceneSynchronizer::NodeData *rew_node_data = scene_synchronizer->get_node_data(*s_snap_it.key);
+		if (rew_node_data == nullptr) {
+			continue;
+		}
+
+		bool recover_this_node = false;
+		const Vector<SceneSynchronizer::VarData> *c_vars = client_snapshots.front().node_vars.lookup_ptr(*s_snap_it.key);
+		if (c_vars == nullptr) {
+			NET_DEBUG_PRINT("Rewind is needed because the client snapshot doesn't contains this node: " + rew_node_data->node->get_path());
```

Comment:
Typo: "doesn't contains this node" -> "doesn't contain this node"

Reply:
Done ty